### ISSUE TITLE
feature request: :warn-on-reflection doesn't warn if there's type hint

### ISSUE THREAD

**Issue by filipesilva** [clj-kondo/clj-kondo#2151]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
I'd like to get reflection warnings without having to set `(set! *warn-on-reflection* true)` everywhere.
If I use `clj-kondo` with the `:warn-on-reflection` linter on, it warns even when there's no clojure reflection warning.

**Describe the solution you'd like**
When I am using `:warn-on-reflection {:level :warning}`, I would like to only get the warning if there's actual reflection going on.

```clj
(def l 23)

;; gets warning
(.toString l)

;; doesn't get warning
(.toString ^Long l)
```

**Describe alternatives you've considered**
Maybe instead of a single warning for the ns, it should be one warning for each reflection instance.
Maybe this should be a flag on `:warn-on-reflection` instead of default behaviour.
Maybe it should be another different linter.
Maybe it shouldn't be done because it's too hard for static analysis.

**Additional context**
https://github.com/jonase/eastwood#reflection seems to work like this. It does not warn if there's no reflection, and it warns once for each reflection warning.

I could do a PR for this with some pointers. If there's a similar functionality already I could use that as an example.

**Comment by borkdude** [comment]
> When I am using :warn-on-reflection {:level :warning}, I would like to only get the warning if there's actual reflection going on.

This is not something clj-kondo can do since it doesn't have access to the runtime of the Clojure program. It is relying solely on static analysis. So it's out of scope. Personally I can live with `(set! *warn-on-reflection* true)` even if I have all the type hints in place.

**Comment by filipesilva** [comment]
Got it, thanks for the fast response!
